# 任务5 图像人脸检测

## 职业能力目标

- 理解HaarCascade级联分类器的使用方法；
- 理解ROI的定义；
- 能够使用CascadeClassifier库完成人脸检测。

## 任务描述

本实验将实现用opencv自带的人脸库对读取的图片进行人脸检测并标注人脸框。

## 任务要求

- 使用imread方法读取图片；
- 使用cvtColor方法转换图像色彩；
- 使用CascadeClassifier方法加载HaarCascade模型检测图片人脸；
- 使用rectangle方法绘制人脸矩形框。

## 任务实施

## 1. 了解人脸检测
人脸检测在日常生活中受到广泛的应用，人脸检测就是在一幅图像中找出所有人脸的位置。

通过OpenCV内置的人脸检测模型，对画面中的人脸绘制出人脸所在位置。

通过人脸检测后再进行人脸识别就可以实现日常生活中应用。例如：扫脸解锁手机，扫脸支付，扫脸开门。

<img src = './src/人脸检测1.png' width = 600 height = 500>

- 人脸变成了一个人非常重要的生物信息。


## 2. 图像人脸检测
读入图片，检测人脸，并在图片中标识人脸所在的矩形区域。

### 2.1 导入cv2

`python-opencv` 在python中的包名称叫做 cv2
- `cv2`实现图像处理和计算机视觉方面的很多通用算法。

In [1]:
import cv2

### 2.2 读入图片
- `cv2.imread(filepath)`：读取彩色图像，图像是按照BGR像素存储

  参数说明：
    - `filepath`要读入图片的完整路径

`img_path`变量定义图片路径，通过`cv2.imread`读取图片，将返回值赋值给`img`

<font color=red size=3>动手练习1</font>

在`<1>`处，请用`cv2.imread()`来读取`img_path`图片。

**填写完成后执行后续代码，`if`语句用于判断图片是否正确读入，若未正确读入图片则`print`输出，反之说明填写正确。**

In [2]:
img_path = './exp/face.jpg'
img = cv2.imread(img_path)

In [3]:
if img is None:
    # 判断图片是否读入正确
    print("ERROR：请检查图片路径")
else:
    print(img)

[[[176 177 181]
  [176 177 181]
  [176 177 181]
  ...
  [174 177 181]
  [174 177 181]
  [173 176 180]]

 [[176 177 181]
  [176 177 181]
  [176 177 181]
  ...
  [174 177 181]
  [174 177 181]
  [173 176 180]]

 [[176 177 181]
  [176 177 181]
  [176 177 181]
  ...
  [174 177 181]
  [174 177 181]
  [173 176 180]]

 ...

 [[180 180 180]
  [182 182 182]
  [184 184 184]
  ...
  [133 138 142]
  [138 141 145]
  [139 142 146]]

 [[183 183 183]
  [184 184 184]
  [186 186 186]
  ...
  [133 138 142]
  [138 141 145]
  [139 142 146]]

 [[183 183 183]
  [184 184 184]
  [186 186 186]
  ...
  [133 138 142]
  [138 141 145]
  [139 142 146]]]


<details>
<summary><font color=red size=3>点击查看动手练习1答案</font></summary>
<pre><code>

```python
# 设置图片路径
img_path = './exp/face.jpg'
img = cv2.imread(img_path)
```
</code></pre>
</details>

### 2.3 图像转换
把`BGR`彩色图像转换成`GRAY`灰度图需要使用`cvtColor`函数进行转换
- `cv2.cvtColor(src, code)`：

  参数说明：
  - `src`：它是要更改其色彩空间的图像。
  - `code`：它是色彩空间转换代码。

<font color=red size=3>动手练习2</font>

在`<1>`处，请用`cv2.cvtColor()`来将彩色图像`img`通过`cv2.COLOR_BGR2GRAY`转化成灰度图`gray`。

**填写完成后执行代码，输出如下所示，说明填写正确。**

```python
gray
```
```
array([[142, 142, 142, ..., 147, 147, 147],
       [142, 142, 142, ..., 147, 147, 147],
       [142, 142, 142, ..., 147, 147, 147],
       ...,
       [112, 112, 113, ..., 110, 110, 111],
       [113, 113, 114, ..., 112, 112, 112],
       [113, 113, 114, ..., 109, 109, 109]], dtype=uint8)

```

In [4]:
# 将彩色图片转换为灰度图
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
gray

array([[178, 178, 178, ..., 178, 178, 177],
       [178, 178, 178, ..., 178, 178, 177],
       [178, 178, 178, ..., 178, 178, 177],
       ...,
       [180, 182, 184, ..., 139, 142, 143],
       [183, 184, 186, ..., 139, 142, 143],
       [183, 184, 186, ..., 139, 142, 143]], dtype=uint8)

<details>
<summary><font color=red size=3>点击查看动手练习2答案</font></summary>
<pre><code>

```python
# 将彩色图片转换为灰度图
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
gray
```
</code></pre>
</details>

### 2.4 载入人脸检测的Cascade模型
`CascadeClassifier`**级联分类器的使用方法**

通过`HaarCascade`模型，输入图片，就可以获取人脸所在区域的矩形位置。

模型的使用方法简单，首先载入对应的`HaarCascade`文件，文件格式为xml。这里已经将文件下载到了`haar`文件夹下， 可以通过相对路径进行引用。
```
├── haar
    └── haarcascade_frontalface_default.xml
```   

<img src = './src/人脸检测3.png'  width = 960></img>


载入人脸检测的Cascade模型，就是在`cv2.CascadeClassifier()`中传入对应`HaarCascade`文件，即`haarcascade_frontalface_default.xml`

`FaceCascade`为分类器对象，以供后续调用使用。

<font color=red size=3>动手练习3</font>

在`<1>`处，请用`cv2.CascadeClassifier`来载入`haar`目录下的`haarcascade_frontalface_default.xml` Cascade人脸检测模型。

**填写完成后执行以下代码，输出结果类似为`<CascadeClassifier 0x7f675be730>`的级联分类器实例对象，说明填写正确。**

In [5]:
# 载入人脸检测的Cascade模型
FaceCascade = cv2.CascadeClassifier('./haar/haarcascade_frontalface_default.xml')
FaceCascade

< cv2.CascadeClassifier 0000020868B72270>

<details>
<summary><font color=red size=3>点击查看动手练习3答案</font></summary>
<pre><code>

```python
# 载入人脸检测的Cascade模型
FaceCascade = cv2.CascadeClassifier('./haar/haarcascade_frontalface_default.xml')
FaceCascade
```
</code></pre>
</details>

**知识补充**

**关于HaarCascade**

OpenCV中人脸识别是通过Haar特征的级联分类器实现，在这里暂不涉及太多底层算法原理部分的讲解。

级联分类器原理说明参考：https://blog.csdn.net/wutao1530663/article/details/78294349

OpenCV里面实际上有很多预先训练好的HaarCascade模型(XML文件) , 例如正脸检测， 眼睛检测， 全身检测，下半身检测等。

<img src = './src/人脸检测2.png'  width = 960>

在OpenCV的两个代码仓库里面都有。

数据1:https://github.com/opencv/opencv/tree/master/data/haarcascades

数据2:https://github.com/opencv/opencv_contrib/tree/246ea8f3bdf174a2aad6216c2601e3a93bf75c29/modules/face/data/cascades


### 2.5 检测画面中的人脸

将图片的灰度图传入到这个`FaceCascade`模型中，进行人脸检测。

- `FaceCascade.detectMultiScale(image, scaleFactor, minNeighbors)`：分类器对象调用参数调节

  参数说明：
  - `image`：输入图像
  - `scaleFactor`：每次缩小图像的比例，默认是1.1
  - `minNeighbors`：匹配成功所需要的周围矩形框的数目，每一个特征匹配到的区域都是一个矩形框，只有多个矩形框同时存在的时候，才认为是匹配成功，比如人脸匹配，默认值是3。


使用`FaceCascade.detectMultiScale`分类器调整参数，输入转换好要检测的灰度图，设置图像比例为1.1，需要有5个矩形框同在才认为匹配成功。

<font color=red size=3>动手练习4</font>

1. 在`<1>`处，请用`FaceCascade.detectMultiScale()`来检测画面中的人脸。传入的参数`image`为`gray`,scaleFactor为`1.1`，`minNeighbors`为5

**填写完成后执行代码，输出结果类似为`array([[ 97, 141, 500, 500]], dtype=int32)`的人脸识别结果区域，说明填写正确。**

In [6]:
# 检测图像中的人脸并返回绘制矩形框的值
faces = FaceCascade.detectMultiScale(image = gray, scaleFactor = 1.1, minNeighbors = 5)
faces

array([[ 40,  63, 208, 208],
       [342,  68, 206, 206]], dtype=int32)

返回值`faces`是人脸所在区域的ROI数组，是识别到的人脸矩形框。例如：[(x1, y1, w1, h1), (x2, y2, w2, h2)]

- 其中`[x, y, w, h]`：

  分别是
    - (x, y): 左上角坐标值
    - w: 人脸矩形区域的宽度
    - h: 人脸矩形区域的高度

<details>
<summary><font color=red size=3>点击查看动手练习4答案</font></summary>
<pre><code>

```python
# 检测图像中的人脸并返回绘制矩形框的值
faces = FaceCascade.detectMultiScale(image = gray, scaleFactor = 1.1, minNeighbors = 5)
faces
```
</code></pre>
</details>

**知识补充**

<img src = './src/ROI.png'  width = 960>

ROI 的全称是Region Of Interest , 用于表示在画面的子区域。整个画面的原点（0,0） ， 在整个画面中的左上角。 ROI 本质上是Tuple类型的数据，其中（x,y）代表人脸所在矩形区域的左上角坐标， w 代表矩形的宽度， h代表矩形的高度。



### 2.6 绘制矩形框

- `cv2.rectangle(img，pt1，pt2，color, thickness)`：这个函数的作用是在图像上绘制一个简单的矩形。

  参数说明：
  - `img`：图片
  - `pt1`：矩形的左上角顶点
  - `pt2`：与pt1相反的矩形的顶点，即右下角顶点
  - `color`：矩形的颜色或亮度（灰度图像）。
  - `thickness`：它是矩形边框线的粗细像素

In [7]:
# 遍历返回的face数组
for face in faces:
    (x, y, w, h) = face
    # 在原彩图上绘制矩形
    cv2.rectangle(img, (x, y), (x+w, y+h), (0, 255, 0), 4)

### 2.7 创建显示窗口

`cv2.namedWindow(winname, flags)`：构建视频的窗口，用于放置图片

参数说明：
- `winname`：表示窗口的名字,可用作窗口标识符的窗口名称。
- `flags`：用于设置窗口的属性，常用属性如下
  - `WINDOW_NORMAL`：可以调整大小窗口
  - `WINDOW_KEEPRATIO`：保持图像比例
  - `WINDOW_GUI_EXPANDED`：绘制一个新的增强GUI窗口

In [8]:
cv2.namedWindow('Face',flags=cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO | cv2.WINDOW_GUI_EXPANDED)
cv2.setWindowProperty('Face', cv2.WND_PROP_FULLSCREEN, cv2.WINDOW_FULLSCREEN) # 全屏展示

### 2.8 显示图片

- `cv2.imshow(winname, mat)`函数可以在窗口中显示图像。

  参数
  - `winname`：窗口名称（也就是我们对话框的名称），它是一个字符串类型。
  - `mat`：是每一帧的画面图像。可以创建任意数量的窗口，但必须使用不同的窗口名称。
         
- `cv2.waitKey`：`waitkey`控制着`imshow`的持续时间，
  - 当`imshow`之后不跟`waitkey`时，相当于没有给`imshow`提供时间展示图像，只会有一个空窗口一闪而过。
  - `cv2.waitKey(100)`表示窗口中显示图像时间为100毫秒
  - `cv2.imshow`之后一定要跟`cv2.waitKey`函数

In [9]:
cv2.imshow('Face', img)
cv2.waitKey(5000)

-1

In [10]:
# 关闭所有的窗口
cv2.destroyAllWindows()

### <font color=red>动手实验</font>

按照以下要求完成实验：
1. 在`<1>`处，读入图片`face1.jpg`，命名为`img`。
2. 在`<2>`处，使用`cv2.cvtColor`函数进行转换，把`BGR`转换成`GRAY`灰度图
3. 在`<3>`处，使用`cv2.CascadeClassifier`传入模型`haarcascade_frontalface_default.xml`
4. 在`<4>`处，使用`FaceCascade.detectMultiScale`检测画面中的人脸，图像缩小的比例是1.1，匹配成功所需要的周围矩形框的数目为3。
5. 在`<5>`处，使用`cv2.rectangle`和循环遍历进行绘制矩形框
6. 在`<6>`处，使用`cv2.namedWindow`创建显示窗口命名为`Face`,属性设置为`可调整大小`，`保持图像比例`
7. 在`<7>`处，使用`cv2.imshow`在窗口`Face`中显示图像`img`
8. 在`<8>`处，设置`cv2.waitKey()`为5000毫秒


**能够成功在图片`face1.jpg`中绘制出矩形框，则表示实验完成。**

In [1]:
import cv2
# 完成代码
#设置图片路径
img_path = './exp/face1.jpg'
#载入带有人脸的图片
img = cv2.imread(img_path)


#将彩色图片转换为灰度图
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)


#载入人脸检测的Cascade模型
FaceCascade = cv2.CascadeClassifier('./haar/haarcascade_frontalface_default.xml')

#检测画面中的人脸
faces = FaceCascade.detectMultiScale(gray, scaleFactor=1.1, minNeighbors=3)


#遍历返回的face数组
for face in faces:
    #解析tuple类型的face位置数据
    #(x, y): 左上角坐标值
    #w: 人脸矩形区域的宽度
    #h: 人脸矩形区域的高度
    (x, y, w, h) = face
    #在原彩图上绘制矩形
    cv2.rectangle(img, (x, y), (x+w, y+h), (0, 255, 0), 4)


#创建一个窗口 名字叫做Face
cv2.namedWindow('Face', flags=cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO)
cv2.setWindowProperty('Face', cv2.WND_PROP_FULLSCREEN, cv2.WINDOW_FULLSCREEN) # 全屏展示


#在窗口Face上面展示图片img
cv2.imshow('Face', img)
#等待任意按键按下
cv2.waitKey(5000)

-1

<details>
<summary><font color=red size=3>点击查看答案</font></summary>
<pre><code>

```python
import cv2

#设置图片路径
img_path = './exp/face1.jpg'
#载入带有人脸的图片
img = cv2.imread(img_path)

#将彩色图片转换为灰度图
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

#载入人脸检测的Cascade模型
FaceCascade = cv2.CascadeClassifier('./haar/haarcascade_frontalface_default.xml')

#检测画面中的人脸
faces = FaceCascade.detectMultiScale(gray, scaleFactor=1.1, minNeighbors=5)

#遍历返回的face数组
for face in faces:
    #解析tuple类型的face位置数据
    #(x, y): 左上角坐标值
    #w: 人脸矩形区域的宽度
    #h: 人脸矩形区域的高度
    (x, y, w, h) = face
    #在原彩图上绘制矩形
    cv2.rectangle(img, (x, y), (x+w, y+h), (0, 255, 0), 4)
    
#创建一个窗口 名字叫做Face
cv2.namedWindow('Face',flags=cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO | cv2.WINDOW_GUI_EXPANDED)
cv2.setWindowProperty('Face', cv2.WND_PROP_FULLSCREEN, cv2.WINDOW_FULLSCREEN) # 全屏展示

#在窗口Face上面展示图片img
cv2.imshow('Face', img)
#等待任意按键按下
cv2.waitKey(1)
```
</code></pre>
</details>

### 2.9 最后释放资源

- `cv2.destroyAllWindows()`：用来删除所有窗口

In [2]:
# 关闭所有的窗口
cv2.destroyAllWindows()

## 任务小结

本次实验的收获：

- 理解了HaarCascade级联分类器的使用方法；
- 理解了ROI的定义；
- 使用cv2自带库完成人脸检测。